core/network/fabric-interop-cc/contracts/interop/manage_assets.go

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright IBM Corp. All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// manage_assets is a chaincode that contains all the code related to asset management operations (e.g., Lock, Unlock, Claim)
+// and any related utility functions
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"crypto/sha256"
+	"encoding/base64"
+	"time"
+
+	"github.com/hyperledger/fabric-contract-api-go/contractapi"
+	"github.com/golang/protobuf/proto"
+	log "github.com/sirupsen/logrus"
+	"github.com/hyperledger-labs/weaver-dlt-interoperability/core/network/fabric-interop-cc/contracts/interop/protos-go/common"
+)
+
+type AssetLockValue struct {
+	Locker	string	`json:"locker"`
+	Recipient	string	`json:"recipient"`
+	Hash	string	`json:"hash"`
+	ExpiryTimeSecs	uint64	`json:"expiryTimeSecs"`
+}
+
+// LockAsset cc is used to record locking of an asset on the ledger
+func (s *SmartContract) LockAsset(ctx contractapi.TransactionContextInterface, assetAgreementBytes string, lockInfoBytes string) error {
+
+	assetAgreement := &common.AssetExchangeAgreement{}
+	err := proto.Unmarshal([]byte(assetAgreementBytes), assetAgreement)
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+	//display the requested asset agreement
+	log.Info(fmt.Printf("assetExchangeAgreement: %+v\n", assetAgreement))
+
+	lockInfoHTLC := &common.AssetLockHTLC{}
+	err = proto.Unmarshal([]byte(lockInfoBytes), lockInfoHTLC)
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+	//display the requested asset agreement
+	log.Info(fmt.Printf("lockInfoHTLC: %+v\n", lockInfoHTLC))
+
+	if lockInfoHTLC.TimeSpec != common.AssetLockHTLC_EPOCH {
+		errorMsg := "only EPOCH time is supported at present"
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	assetLockKey := assetAgreement.Type + ":" + assetAgreement.Id
```

Comment:
I wonder if the locker also ought to be in the key? I know we are considering for now that an asset of a given type has a unique ID within a network, but perhaps that assumption may not always hold up?

Reply:
Where did we reach this conclusion? I still believe the locker ought to be part of the key (this will help with indexing when you implement list queries). In fact, the contract ID is also needed, I think. When you implement fungible assets, just using asset type and number of units won't suffice for the key. You will need the contract ID to generate a unique ledger key for the asset lock.